File: day15/day15a.py
#!/usr/bin/python3

with open('day15.txt', 'r') as f:
    nums = [int(s) for s in f.read().strip().split(',')]
    while len(nums) < 2020:
        num = nums[-1]
        # If the most recent number has not been spoken before, say 0
        next_num = 0
        for i in range(1, len(nums)):
            # Search backwards to see whether the most recent number has been spoken before
            j = len(nums) - 1 - i
            if nums[j] == num:
                # Specify how many turns apart the previous occurrence was
                next_num = i
                break
        nums.append(next_num)
    print(nums[-1])

File: day15/day15a.cpp
#include <fstream>
#include <iostream>
#include <string>
#include <vector>

int main() {
  // Read input from file
  std::ifstream file("day15.txt");
  std::string line;
  std::vector<int> nums;

  while(std::getline(file, line, ',')) {
    int next_int = std::stoi(line);
    nums.push_back(next_int);
  }

  while(nums.size() < 2020) {
    int num = nums[nums.size() - 1];
    int next_num = 0; // If most recent number has not been spoken before, say 0 by default

    // Search backwards to see whether most recent number has been spoken before
    for(int i = 1; i < nums.size(); i++) {
      if(nums[nums.size() - 1 - i] == num) {
        // If most recent number was spoken before, specify how many turns apart the previous occurrence was
        next_num = i;
        break;
      }
    }
    
    nums.push_back(next_num);
  }

  std::cout << nums[nums.size() - 1] << std::endl;
  return 0;
}
